Adding standard metrics

hectorhdzg opened the pull request:
Added better folder structure to the project

Fixes #41

hectorhdzg committed:
WIP

hectorhdzg committed:
Merge branch 'master' of https://github.com/microsoft/opentelemetry-exporters-python

hectorhdzg committed:
Added tests

hectorhdzg committed:
Fixing lint

hectorhdzg committed:
Fixing lint for real

hectorhdzg committed:
Fixing python 3.5 issue

hectorhdzg committed:
Adding test clean up back

hectorhdzg commented:
Not sure why CI is being executed twice here, @lzchen any idea?

hectorhdzg committed:
Adding import shutil

lzchen commented:
@hectorhdzg 
See [this](https://github.com/checkstyle/sonar-checkstyle/issues/261).
Two builds are being triggered because you are making a PR from a branch on origin, not on your fork.

lzchen commented on tox.ini:
```diff
@@ -34,20 +34,18 @@ commands =
 
 commands_pre =
-  pip install {toxinidir}/azure_monitor
+  python scripts/eachdist.py install --editable
```
Do we need an `eachdist.py` file? Our packages are not as complicated so there's not really a need to have a separate script to run or CI. Also, I'm not sure you can simply copy the script from OT, maybe there are legal implications?

hectorhdzg replied:
Current bash solution requires contributors to install multiple tools to have it working in Windows, python execution is multi platform friendly, I have no idea about legal implications, we are taking a lot of code from OC and OT even if we wrote it in the first place, so we better understand this better

hectorhdzg replied:
Added bash in the meantime, will follow up about this

lzchen commented on azure_monitor/src/azure_monitor/auto_collection/__init__.py:
```diff
@@ -0,0 +1,36 @@
+]
+
+
+class AutoCollection:
```
I don't see `AutoCollection` being used anywhere?

hectorhdzg replied:
Currently user need to instantiate it, once wrapper is ready we can do it depending on configuration

lzchen commented on azure_monitor/src/azure_monitor/auto_collection/__init__.py:
```diff
@@ -0,0 +1,36 @@
+
+
+class AutoCollection:
+    def __init__(
```
An example would also be nice.

lzchen commented on azure_monitor/src/azure_monitor/auto_collection/__init__.py:
```diff
@@ -0,0 +1,36 @@
+        self._request_metrics = RequestMetrics(meter, label_set)
+
+        self._collect_task = PeriodicTask(
+            interval=collection_interval, function=self.collect
```
Should we really be having a separate worker thread for standard metrics?

hectorhdzg replied:
We don't have any other worker thread, and this need to be periodically retrieved so I don't see how else we can do this, any specific idea you have in mind?

lzchen replied:
This leads me to believe that autocollection might have to happen on the sdk rather than in the exporter. It might need to be a separate `azure` owned wrapper around the OT sdk. We should bring this up in the SIG/standup.

lzchen commented on azure_monitor/src/azure_monitor/auto_collection/dependency_metrics.py:
```diff
@@ -0,0 +1,72 @@
+            "Outgoing Requests per second",
+            "rps",
+            int,
+            Gauge,
```
With OT's [removal of gauge](https://github.com/open-telemetry/opentelemetry-python/issues/403), we will have to refactor this after the beta release. Do you still want to merge with these changes?

hectorhdzg replied:
Yes, refactor PR to align with OT beta code would be super clear instead of having huge PRs like this one

lzchen commented on azure_monitor/src/azure_monitor/auto_collection/dependency_metrics.py:
```diff
@@ -0,0 +1,72 @@
+    def _track_dependency_rate(self) -> None:
+        """ Track Dependency rate
+
+        Calculated by obtaining by getting the number of outgoing requests made
```
```suggestion
        Calculated by obtaining the number of outgoing requests made
```

lzchen commented on azure_monitor/src/azure_monitor/auto_collection/request_metrics.py:
```diff
@@ -0,0 +1,142 @@
+
+    def track(self) -> None:
+        self._track_request_duration()
+
```
Missing `_track_request_rate()`?

lzchen commented on azure_monitor/src/azure_monitor/auto_collection/request_metrics.py:
```diff
@@ -0,0 +1,142 @@
+            int,
+            Gauge,
+        )
+        self._request_duration_handle = request_duration_metric.get_handle(
```
Handles are also [removed](https://github.com/open-telemetry/opentelemetry-python/issues/468). This will be quite a bit of refactor work.

hectorhdzg replied:
get_handle= bind it would be an easy change, I believe having OT Beta specific changes in a single PR would help us understand if we are missing something

hectorhdzg committed:
Adding auto collection example

lzchen reviewed (changes requested):
Couple of comments. I think we should separate this PR out from standard metrics changes and CI fixing changes, mostly because a finalized metrics API isn't released yet for OT, and also you can unblock the CI for others with your lint changes.

hectorhdzg committed:
Adding AutoCollection export

hectorhdzg committed:
Fixing lint

hectorhdzg committed:
Update azure_monitor/src/azure_monitor/auto_collection/dependency_metrics.py

Co-Authored-By: Leighton Chen <lechen@microsoft.com>

hectorhdzg commented:
@lzchen thanks for reviewing, I believe I addressed most of your comments, it would be quite overhead to split this in 2 PRs lint and standard metrics, if you have strong opinions about it let me know.

lzchen reviewed (approved):
LGTM

hectorhdzg committed:
Removing eachdist.ini

hectorhdzg committed:
Merge branch 'hectorh/standardMetrics' of https://github.com/microsoft/opentelemetry-exporters-python into hectorh/standardMetrics

lzchen commented on scripts/pylint.sh:
```diff
@@ -10,4 +10,4 @@ function pylint_dir {
 }
 
-pylint_dir
+pylint_dir
```
new line here

hectorhdzg committed:
Adding missing space

hectorhdzg committed:
Updating tox

hectorhdzg committed:
Fixing lint
